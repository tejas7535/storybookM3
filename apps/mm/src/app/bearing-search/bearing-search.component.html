<mat-form-field class="example-full-width">
  <mat-label>Bearing</mat-label>
  <input
    type="text"
    placeholder="Search for SCHAEFFLER bearings"
    aria-label="text"
    matInput
    [formControl]="myControl"
    [matAutocomplete]="auto"
  />
  <mat-autocomplete
    #auto="matAutocomplete"
    [displayWith]="displayValue"
    (optionSelected)="selectBearing($event)"
  >
    <mat-option *ngFor="let option of options$ | ngrxPush" [value]="option">
      {{ option.title }}
    </mat-option>
  </mat-autocomplete>
</mat-form-field>
