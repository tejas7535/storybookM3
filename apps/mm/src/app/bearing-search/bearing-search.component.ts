import { HttpClient } from '@angular/common/http';
import {
  ChangeDetectionStrategy,
  Component,
  EventEmitter,
  OnInit,
  Output,
} from '@angular/core';
import { FormControl } from '@angular/forms';
import { MatAutocompleteSelectedEvent } from '@angular/material/autocomplete';

import { Observable, of } from 'rxjs';
import {
  debounceTime,
  filter,
  map,
  startWith,
  switchMap,
} from 'rxjs/operators';

import { environment } from '../../environments/environment';

interface BearingOption {
  id: string;
  title: string;
}

interface SearchEntry {
  data: BearingOption;
  links: [];
  _media?: any;
}

interface SearchResult {
  data: SearchEntry[];
}

@Component({
  selector: 'mm-bearing-search',
  templateUrl: './bearing-search.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class BearingSearchComponent implements OnInit {
  @Output() bearing = new EventEmitter<string | undefined>();

  myControl = new FormControl();

  options$: Observable<BearingOption[]> = of([]);

  constructor(private readonly http: HttpClient) {}

  ngOnInit(): void {
    this.options$ = this.myControl.valueChanges.pipe(
      startWith(''),
      debounceTime(300),
      filter((value) => value.length > 1),
      switchMap((value) => this.getBearings(value))
    );
  }

  getBearings(searchQuery: string): Observable<BearingOption[]> {
    const requestUrl = `${environment.apiMMBaseUrl}/bearing/search/?pattern=${searchQuery}&page=1&size=1000`;

    return this.http.get<SearchResult>(requestUrl).pipe(
      map((response) =>
        response.data.map((entry: SearchEntry) => {
          const { title, id } = entry.data;

          return { title, id };
        })
      )
    );
  }

  displayValue(option: BearingOption): string {
    return option && option.title ? option.title : '';
  }

  selectBearing(event: MatAutocompleteSelectedEvent): void {
    this.bearing.emit(event.option.value.id);
  }
}
