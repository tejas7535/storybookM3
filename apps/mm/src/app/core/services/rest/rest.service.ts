import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';

import { map, Observable } from 'rxjs';

import { TranslocoService } from '@jsverse/transloco';
import { CalculationRequestPayload } from '@mm/shared/models/calculation-request.model';
import { withCache } from '@ngneat/cashew';

import { environment } from '../../../../environments/environment';
import {
  MMBearingPreflightResponse,
  PreflightRequestBody,
  SearchResult,
  ShaftMaterialResponse,
} from '../../../shared/models';
import { BearinxOnlineResult } from '../bearinx-result.interface';

@Injectable({
  providedIn: 'root',
})
export class RestService {
  private readonly httpClient = inject(HttpClient);
  private readonly translocoService = inject(TranslocoService);

  private readonly bearingCalculationPath = `${environment.baseUrl}/calculate`;

  public getBearingSearch(searchQuery: string): Observable<SearchResult> {
    return this.httpClient.get<SearchResult>(
      `${environment.baseUrl}/bearings/search?pattern=${searchQuery}&page=1&size=1000`
    );
  }

  public getBearingCalculationResult(
    requestPayload: CalculationRequestPayload
  ): Observable<BearinxOnlineResult> {
    return this.httpClient.post<BearinxOnlineResult>(
      this.bearingCalculationPath,
      requestPayload,
      {
        context: withCache({ version: JSON.stringify(requestPayload) }),
      }
    );
  }

  public getBearingPreflightResponse(
    body: PreflightRequestBody
  ): Observable<MMBearingPreflightResponse> {
    const path = `${environment.baseUrl}/dialog`;

    return this.httpClient.post<MMBearingPreflightResponse>(path, body);
  }

  public getBearingsMaterialResponse(
    idmmShaftMaterial: string
  ): Observable<ShaftMaterialResponse> {
    return this.httpClient.get<ShaftMaterialResponse>(
      `${environment.baseUrl}/materials/${idmmShaftMaterial}`,
      { context: withCache({ version: this.translocoService.getActiveLang() }) }
    );
  }

  public getLoadOptions<T>(requestUrl: string): Observable<T> {
    return this.httpClient.get<T>(requestUrl, {
      context: withCache({ version: this.translocoService.getActiveLang() }),
    });
  }

  public getBearinxVersions(): Observable<{ [key: string]: string }> {
    return this.httpClient
      .get<
        { name: string; version: string }[]
      >(`${environment.bearinxApiBaseUrl}/version`)
      .pipe(
        map((response) =>
          Object.fromEntries(
            response.map(({ name, version }) => [name, version])
          )
        )
      );
  }
}
