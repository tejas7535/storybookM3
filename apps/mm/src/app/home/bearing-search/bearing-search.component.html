<schaeffler-search
  *transloco="let t"
  appearance="outline"
  [stringOptions]="bearingResultList$ | ngrxPush"
  [loading]="bearingSelectionLoading$ | ngrxPush"
  [control]="myControl"
  [label]="t('form.chooseOption')"
  [placeholder]="t('form.chooseOption')"
  [noResultsText]="t('form.searchNoResults')"
  [initialValue]="selectedBearing"
  (searchUpdated)="getBearings($event)"
  (optionSelected)="onOptionSelected($event)"
>
  <div loadingContent>{{ t('form.searchLoading') }}</div>
</schaeffler-search>
