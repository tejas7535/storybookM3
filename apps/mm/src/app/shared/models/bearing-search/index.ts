export * from './bearing-search.model';
