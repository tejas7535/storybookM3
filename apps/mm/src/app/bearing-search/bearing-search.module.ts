import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatInputModule } from '@angular/material/input';

import { ReactiveComponentModule } from '@ngrx/component';

import { BearingSearchComponent } from './bearing-search.component';

@NgModule({
  declarations: [BearingSearchComponent],
  imports: [
    CommonModule,

    // NGRX
    ReactiveComponentModule,

    // Forms
    ReactiveFormsModule,

    // Angular Material
    MatAutocompleteModule,
    MatInputModule,
  ],
  exports: [BearingSearchComponent],
})
export class BearingSearchModule {}
