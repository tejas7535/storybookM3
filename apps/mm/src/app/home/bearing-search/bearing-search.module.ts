import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { SearchAutocompleteModule } from '@schaeffler/search-autocomplete';

import { SharedModule } from './../../shared/shared.module';
import { BearingSearchComponent } from './bearing-search.component';

@NgModule({
  declarations: [BearingSearchComponent],
  imports: [CommonModule, SharedModule, SearchAutocompleteModule],
  exports: [BearingSearchComponent],
})
export class BearingSearchModule {}
