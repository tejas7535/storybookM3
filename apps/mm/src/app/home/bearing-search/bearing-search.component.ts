import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  EventEmitter,
  Input,
  Output,
} from '@angular/core';
import { UntypedFormControl } from '@angular/forms';

import { map, Observable } from 'rxjs';

import { CalculationSelectionFacade } from '@mm/core/store/facades/calculation-selection/calculation-selection.facade';
import { PushPipe } from '@ngrx/component';

import { ApplicationInsightsService } from '@schaeffler/application-insights';
import { StringOption } from '@schaeffler/inputs';
import { SearchModule } from '@schaeffler/inputs/search';
import { SharedTranslocoModule } from '@schaeffler/transloco';

import { BEARING } from '../../shared/constants/tracking-names';
import { BearingOption } from '../../shared/models';
@Component({
  standalone: true,
  imports: [SearchModule, PushPipe, SharedTranslocoModule],
  selector: 'mm-bearing-search',
  templateUrl: './bearing-search.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class BearingSearchComponent {
  @Input() public selectedBearing?: BearingOption;
  @Output() public bearing = new EventEmitter<string | undefined>();

  public myControl = new UntypedFormControl('');

  public bearingSelectionLoading$ = this.calculationSelectionFacade.isLoading$;

  public bearingResultList$: Observable<StringOption[]> =
    this.calculationSelectionFacade.bearingResultList$.pipe(
      map((bearingDesignation) => {
        if (!bearingDesignation) {
          return [];
        }
        const options: StringOption[] = bearingDesignation.map((result) => ({
          id: result.id,
          title: result.title,
        }));

        return options;
      })
    );

  private readonly minimumChars = 2;

  public constructor(
    private readonly calculationSelectionFacade: CalculationSelectionFacade,
    private readonly applicationInsightsService: ApplicationInsightsService,
    private readonly changeDetector: ChangeDetectorRef
  ) {}

  public getBearings(searchQuery: string): void {
    if (searchQuery?.length >= this.minimumChars) {
      this.calculationSelectionFacade.searchBearing(searchQuery);
      this.changeDetector.detectChanges();
    } else {
      this.calculationSelectionFacade.resetBearingSelection();
    }
  }

  public onOptionSelected(selection: StringOption): void {
    if (selection) {
      const { id, title } = selection;
      this.trackBearingSelection(title, id as string);
      this.bearing.emit(id as string);
    }
  }

  public trackBearingSelection(bearing: string, selectionId: string): void {
    this.applicationInsightsService.logEvent(BEARING, {
      name: bearing,
      id: selectionId,
    });
  }
}
