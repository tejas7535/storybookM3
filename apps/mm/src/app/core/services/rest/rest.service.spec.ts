import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';

import { createServiceFactory, SpectatorService } from '@ngneat/spectator/jest';

import { ENV_CONFIG } from '@schaeffler/http';

import {
  MMBearingPreflightResponse,
  MMBearingsMaterialResponse,
  MMResponseVariants,
  PreflightRequestBody,
} from '../../../shared/models';
import { environment } from './../../../../environments/environment';
import {
  BEARING_CALCULATION_RESULT_MOCK,
  BEARING_MATERIAL_RESPONSE_MOCK,
  BEARING_PREFLIGHT_RESPONSE_MOCK,
  BEARING_SEARCH_RESULT_MOCK,
  LOAD_OPTIONS_RESPONSE_MOCK,
} from './../../../../testing/mocks/rest.service.mock';
import { SearchResult } from './../../../shared/models/bearing-search/bearing-search.model';
import { RestService } from './rest.service';

describe('RestService', () => {
  let spectator: SpectatorService<RestService>;
  let service: RestService;
  let httpMock: HttpTestingController;

  const createService = createServiceFactory({
    service: RestService,
    imports: [HttpClientTestingModule],
    providers: [
      RestService,
      {
        provide: ENV_CONFIG,
        useValue: {
          environment: {
            baseUrl: '',
            preflightPath: 'bearing-preflight',
            materialsPath: 'materialdata/id/',
            bearingRelationsPath: 'bearing-relations/',
            bearingCalculationPath: 'bearing-calculation',
          },
        },
      },
    ],
  });

  beforeEach(() => {
    spectator = createService();
    service = spectator.inject(RestService);
    httpMock = spectator.inject(HttpTestingController);
  });

  it('should be created', () => {
    expect(service).toBeTruthy();
  });

  describe('#getBearingSearch', () => {
    it('should send a bearing search request with given query', (done) => {
      service.getBearingSearch('theQuery').subscribe((result: SearchResult) => {
        expect(result).toEqual(BEARING_SEARCH_RESULT_MOCK);
        done();
      });

      const req = httpMock.expectOne(
        `/bearing/search/?pattern=theQuery&page=1&size=1000`
      );
      expect(req.request.method).toBe('GET');
      req.flush(BEARING_SEARCH_RESULT_MOCK);
    });
  });

  describe('#getBearingRelations', () => {
    it('should send a bearing relation request with given id', (done) => {
      service.getBearingRelations('theId').subscribe((result: SearchResult) => {
        expect(result).toEqual({});
        done();
      });

      const req = httpMock.expectOne(
        `/${environment.bearingRelationsPath}theId`
      );
      expect(req.request.method).toBe('GET');
      req.flush({});
    });
  });

  describe('#getBearingCalculationResult', () => {
    it('should send a bearing calculation request with given form properties', (done) => {
      service
        .getBearingCalculationResult('some form properties i guess')
        .subscribe((result: SearchResult) => {
          expect(result).toEqual(BEARING_CALCULATION_RESULT_MOCK);
          done();
        });

      const req = httpMock.expectOne(`/${environment.bearingCalculationPath}`);
      expect(req.request.method).toBe('POST');
      req.flush(BEARING_CALCULATION_RESULT_MOCK);
    });
  });

  describe('#getBearingPreflightResponse', () => {
    it('should send a bearing preflight request with given body', (done) => {
      service
        .getBearingPreflightResponse({} as PreflightRequestBody)
        .subscribe((result: MMBearingPreflightResponse) => {
          expect(result).toEqual(BEARING_PREFLIGHT_RESPONSE_MOCK);
          done();
        });

      const req = httpMock.expectOne(`/${environment.preflightPath}`);
      expect(req.request.method).toBe('POST');
      req.flush(BEARING_PREFLIGHT_RESPONSE_MOCK);
    });
  });

  describe('#getBearingMaterialResponse', () => {
    it('should send a bearing material request with given material', (done) => {
      const mockShaftMaterial = 'material1';
      service
        .getBearingsMaterialResponse(mockShaftMaterial)
        .subscribe((result: MMBearingsMaterialResponse) => {
          expect(result).toEqual(BEARING_MATERIAL_RESPONSE_MOCK);
          done();
        });

      const req = httpMock.expectOne(
        `/${environment.materialsPath}${mockShaftMaterial}?cache$=true`
      );
      expect(req.request.method).toBe('GET');
      req.flush(BEARING_MATERIAL_RESPONSE_MOCK);
    });
  });

  describe('#getLoadOptions', () => {
    it('should send a load option request with given url', (done) => {
      service.getLoadOptions('aUrl').subscribe((result: MMResponseVariants) => {
        expect(result).toEqual(LOAD_OPTIONS_RESPONSE_MOCK);
        done();
      });

      const req = httpMock.expectOne('/aUrl?cache$=true');
      expect(req.request.method).toBe('GET');
      req.flush(LOAD_OPTIONS_RESPONSE_MOCK);
    });
  });
});
