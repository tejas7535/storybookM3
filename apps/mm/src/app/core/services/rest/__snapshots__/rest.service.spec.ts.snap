// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestService #fetchBearingInfo should fetch bearing details and map them correctly 1`] = `
Object {
  "bearingId": "ABC123",
  "isHydraulic": true,
  "isMechanical": false,
  "isThermal": true,
  "title": "Test Bearing",
}
`;

exports[`RestService #fetchBearingInfo should handle missing properties with fallback values 1`] = `
Object {
  "bearingId": "DEF456",
  "isHydraulic": false,
  "isMechanical": true,
  "isThermal": false,
  "title": "Partial Bearing",
}
`;

exports[`RestService #fetchBearingInfo should handle missing title with fallback to empty string 1`] = `
Object {
  "bearingId": "GHI789",
  "isHydraulic": false,
  "isMechanical": true,
  "isThermal": false,
  "title": "",
}
`;

exports[`RestService #getBearingSeats should fetch bearing seats and map them correctly 1`] = `
Array [
  Object {
    "id": "seat1",
    "imageUrl": "https://bearinx-d.schaeffler.com/mounting/api/images/seat1.png",
    "text": "Bearing Seat 1",
  },
  Object {
    "id": "seat2",
    "imageUrl": undefined,
    "text": "Bearing Seat 2",
  },
]
`;

exports[`RestService #getMeasurementMethods should fetch measurement methods and map them correctly 1`] = `
Array [
  Object {
    "id": "method1",
    "imageUrl": "https://bearinx-d.schaeffler.com/mounting/api/images/method1.png",
    "text": "Measurement Method 1",
  },
  Object {
    "id": "method2",
    "imageUrl": undefined,
    "text": "Measurement Method 2",
  },
]
`;

exports[`RestService #getNonThermalBearingMountingMethods should fetch non-thermal bearing mounting methods and map them correctly 1`] = `
Array [
  Object {
    "id": "mount1",
    "imageUrl": "https://bearinx-d.schaeffler.com/mounting/api/images/mount1.png",
    "text": "Mounting Method 1",
  },
  Object {
    "id": "mount2",
    "imageUrl": undefined,
    "text": "Mounting Method 2",
  },
]
`;

exports[`RestService #getThermalBearingMountingMethods should fetch thermal bearing mounting methods and map them correctly 1`] = `
Array [
  Object {
    "id": "thermal1",
    "imageUrl": "https://bearinx-d.schaeffler.com/mounting/api/images/thermal1.png",
    "text": "Thermal Mount 1",
  },
  Object {
    "id": "thermal2",
    "imageUrl": undefined,
    "text": "Thermal Mount 2",
  },
]
`;

exports[`RestService #searchBearings should filter out bearings with all flags set to false 1`] = `
Array [
  Object {
    "id": "Valid-Bearing-1",
    "isHydraulic": false,
    "isMechanical": false,
    "isThermal": true,
    "title": "Valid-Bearing-1",
  },
  Object {
    "id": "Valid-Bearing-2",
    "isHydraulic": false,
    "isMechanical": true,
    "isThermal": false,
    "title": "Valid-Bearing-2",
  },
]
`;

exports[`RestService #searchBearings should search bearings and map response to BearingOption array 1`] = `
Array [
  Object {
    "id": "Bearing-1",
    "isHydraulic": true,
    "isMechanical": false,
    "isThermal": true,
    "title": "Bearing-1",
  },
  Object {
    "id": "Bearing-2",
    "isHydraulic": false,
    "isMechanical": true,
    "isThermal": false,
    "title": "Bearing-2",
  },
]
`;

exports[`RestService getBearinxVersions should call the service to get bearinx versions 1`] = `
Object {
  "bearinx": "1",
}
`;
